### ISSUE TITLE
Revocation of certificate failed after upgrading to 8.2.0 - WFLYEJB0034: Jakarta Enterprise Beans Invocation failed

### ISSUE THREAD

**Issue by 3keyroman** [Keyfactor/ejbca-ce#510]
### Discussed in https://github.com/Keyfactor/ejbca-ce/discussions/494

<div type='discussions-op-text'>

<sup>Originally posted by **3keyroman** February  6, 2024</sup>
Hello Community,

I am having an issue revoking some certificates after the upgrade to EJBCA 8.2.0. Trying to revoke certificate from `adminweb` or through the `WS` gives the following exception:

```
2024-02-06 11:35:15,939 ERROR [org.jboss.as.ejb3.invocation] (default task-1) WFLYEJB0034: Jakarta Enterprise Beans Invocation failed on component CertificateStoreSessionBean for method public abstract boolean org.cesecore.certificates.certificate.CertificateStoreSessionLocal.setRevokeStatus(org.cesecore.authentication.tokens.AuthenticationToken,org.cesecore.certificates.certificate.CertificateDataWrapper,java.util.Date,java.util.Date,int) throws org.cesecore.certificates.certificate.CertificateRevokeException,org.cesecore.authorization.AuthorizationDeniedException: javax.ejb.EJBTransactionRolledbackException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [org.cesecore.certificates.certificate.CertificateData#107c124d5c8ecdbee08538fb19aaf213cd1fda9c]
	at org.jboss.as.ejb3@26.1.2.Final//org.jboss.as.ejb3.tx.CMTTxInterceptor.invokeInCallerTx(CMTTxInterceptor.java:219)
	at org.jboss.as.ejb3@26.1.2.Final//org.jboss.as.ejb3.tx.CMTTxInterceptor.required(CMTTxInterceptor.java:392)
	at org.jboss.as.ejb3@26.1.2.Final//org.jboss.as.ejb3.tx.CMTTxInterceptor.processInvocation(CMTTxInterceptor.java:160)
	at org.jboss.invocation@1.7.0.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
	at org.jboss.invocation@1.7.0.Final//org.jboss.invocation.InterceptorContext$Invocation.proceed(InterceptorContext.java:509)
	at org.jboss.weld.core@3.1.9.Final//org.jboss.weld.module.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:72)
	at org.jboss.as.weld.common@26.1.2.Final//org.jboss.as.weld.ejb.EjbRequestScopeActivationInterceptor.processInvocation(EjbRequestScopeActivationInterceptor.java:89)
	at org.jboss.invocation@1.7.0.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
	at org.jboss.as.ejb3@26.1.2.Final//org.jboss.as.ejb3.component.interceptors.CurrentInvocationContextInterceptor.processInvocation(CurrentInvocationContextInterceptor.java:41)
	at org.jboss.invocation@1.7.0.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
	at org.jboss.as.ejb3@26.1.2.Final//org.jboss.as.ejb3.component.invocationmetrics.WaitTimeInterceptor.processInvocation(WaitTimeInterceptor.java:47)
	at org.jboss.invocation@1.7.0.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
	at org.jboss.as.ejb3@26.1.2.Final//org.jboss.as.ejb3.security.IdentityOutflowInterceptor.processInvocation(IdentityOutflowInterceptor.java:73)
	at org.jboss.invocation@1.7.0.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
	at org.jboss.as.ejb3@26.1.2.Final//org.jboss.as.ejb3.security.SecurityDomainInterceptor.processInvocation(SecurityDomainInterceptor.java:44)
	at org.jboss.invocation@1.7.0.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
	at org.jboss.as.ejb3@26.1.2.Final//org.jboss.as.ejb3.deployment.processors.StartupAwaitInterceptor.processInvocation(StartupAwaitInterceptor.java:22)
	at org.jboss.invocation@1.7.0.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
	at org.jboss.as.ejb3@26.1.2.Final//org.jboss.as.ejb3.component.interceptors.ShutDownInterceptorFactory$1.processInvocation(ShutDownInterceptorFactory.java:64)
	at org.jboss.invocation@1.7.0.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
	at org.jboss.as.ejb3@26.1.2.Final//org.jboss.as.ejb3.component.interceptors.LoggingInterceptor.processInvocation(LoggingInterceptor.java:67)
	at org.jboss.invocation@1.7.0.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
	at org.jboss.as.ee@26.1.2.Final//org.jboss.as.ee.component.NamespaceContextInterceptor.processInvocation(NamespaceContextInterceptor.java:50)
	at org.jboss.invocation@1.7.0.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
	at org.jboss.invocation@1.7.0.Final//org.jboss.invocation.ContextClassLoaderInterceptor.processInvocation(ContextClassLoaderInterceptor.java:60)
	at org.jboss.invocation@1.7.0.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
	at org.jboss.invocation@1.7.0.Final//org.jboss.invocation.InterceptorContext.run(InterceptorContext.java:438)
	at org.wildfly.security.elytron-base@1.19.1.Final//org.wildfly.security.manager.WildFlySecurityManager.doChecked(WildFlySecurityManager.java:633)
	at org.jboss.invocation@1.7.0.Final//org.jboss.invocation.AccessCheckingInterceptor.processInvocation(AccessCheckingInterceptor.java:57)
	at org.jboss.invocation@1.7.0.Final//org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:422)
	at org.jboss.invocation@1.7.0.Final//org.jboss.invocation.ChainedInterceptor.processInvocation(ChainedInterceptor.java:53)
	at org.jboss.as.ee@26.1.2.Final//org.jboss.as.ee.component.ViewService$View.invoke(ViewService.java:198)
	at org.jboss.as.ee@26.1.2.Final//org.jboss.as.ee.component.ViewDescription$1.processInvocation(ViewDescription.java:191)
	at org.jboss.as.ee@26.1.2.Final//org.jboss.as.ee.component.ProxyInvocationHandler.invoke(ProxyInvocationHandler.java:81)
	at deployment.ejbca.ear.cesecore-ejb.jar//org.cesecore.certificates.certificate.CertificateStoreSessionLocal$$$view195.setRevokeStatus(Unknown Source)
	at deployment.ejbca.ear.cesecore-ejb.jar//org.cesecore.certificates.certificate.NoConflictCertificateStoreSessionBean.setRevokeStatus(NoConflictCertificateStoreSessionBean.java:298)
...
Caused by: javax.persistence.OptimisticLockException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [org.cesecore.certificates.certificate.CertificateData#107c124d5c8ecdbee08538fb19aaf213cd1fda9c]
	at org.hibernate@5.3.28.Final//org.hibernate.internal.ExceptionConverterImpl.wrapStaleStateException(ExceptionConverterImpl.java:223)
	at org.hibernate@5.3.28.Final//org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:93)
	at org.hibernate@5.3.28.Final//org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181)
	at org.hibernate@5.3.28.Final//org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:188)
	at org.hibernate@5.3.28.Final//org.hibernate.internal.SessionImpl.fireMerge(SessionImpl.java:917)
	at org.hibernate@5.3.28.Final//org.hibernate.internal.SessionImpl.merge(SessionImpl.java:891)
	at org.jboss.as.jpa@26.1.2.Final//org.jboss.as.jpa.container.AbstractEntityManager.merge(AbstractEntityManager.java:568)
	at deployment.ejbca.ear.cesecore-ejb.jar//org.cesecore.certificates.certificate.CertificateStoreSessionBean.setRevokeStatusNoAuth(CertificateStoreSessionBean.java:1327)
	at deployment.ejbca.ear.cesecore-ejb.jar//org.cesecore.certificates.certificate.CertificateStoreSessionBean.setRevokeStatus(CertificateStoreSessionBean.java:1232)
...
Caused by: org.hibernate.StaleObjectStateException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [org.cesecore.certificates.certificate.CertificateData#107c124d5c8ecdbee08538fb19aaf213cd1fda9c]
	at org.hibernate@5.3.28.Final//org.hibernate.event.internal.DefaultMergeEventListener.entityIsDetached(DefaultMergeEventListener.java:356)
	at org.hibernate@5.3.28.Final//org.hibernate.event.internal.DefaultMergeEventListener.onMerge(DefaultMergeEventListener.java:188)
	at org.hibernate@5.3.28.Final//org.hibernate.event.internal.DefaultMergeEventListener.onMerge(DefaultMergeEventListener.java:72)
	at org.hibernate@5.3.28.Final//org.hibernate.internal.SessionImpl.fireMerge(SessionImpl.java:905)
	... 353 more
```

This happens on some of the certificates, not all of them. Looking into the database, there are no obvious differences between the certificates that can be revoked, and certificates that will throw an exception when trying to revoke.

Moreover, the certificate that failed to be revoked through the `adminweb` or `WS`, can be revoked without any issues through the `raweb`. I assume that is because they use different methods.

Does anyone have or had similar issue? Can you advise what can be an issue here?

I am running with database MariaDB Server version: 10.3.39-MariaDB-0+deb10u1 Debian 10, with OpenJDK 17, and on the Wildfly 26.1.2.Final.

I was trying to downgrade to OpenJDK 11, but the problem persists.

With best regards,
Roman</div>

**Comment by primetomas** [comment]
A workaround should be something like:
update CertificateData set invalidityDate=-1 where invalidityDate is null;
Since database protection is versioned it will not break existing signatures, i.e. new fields are not included in old signatures which is known at verification.

**Comment by primetomas** [comment]
Fixed for EJBCA 8.3.0. Should only affect Enterprise, since it depends on database protection. Workaround noted above.

Thanks for the report @3keyroman. Very very helpful.
